### ISSUE TITLE
can not open pdf 

### ISSUE THREAD

**Issue by frankyifei** [koreader/koreader-base#421]
get this error on my kindle and there is no problem on emulator
`# cannot open document /mnt/us/documents/1.pdf ./ffi/koptcontext.lua:13 /lib/libc.so.6 version 'GLIBC_2.22' not found (required by libs/liblept.so.3)`
compiled with GCC 5.3 shipped with ubuntu 16.04, and I dont know if this is the reason

**Comment by NiLuJe** [comment]
Yup, the TC used is the issue.

You either need a TC targeting the same Kernel/Glibc as the target (ideally roughly the same STDC++ implementation, too, but since that's a huge hassle, we link that statically, so, forget about it), or employ a crapload of wrokarounds to avoid pulling in newer symbols.

**Comment by NiLuJe** [comment]
If you want to dig into the "workarounds" approach, search for what those exact GLIBC_2.22 symbols refer to in libs/liblept.so.3 w/ readelf

**Comment by NiLuJe** [comment]
If your Kindle is running FW >= 5.6.5, things become slightly less messy because they updated to a not so ancient glibc, 2.19.

**Comment by frankyifei** [comment]
can I link glibc statically？

**Comment by NiLuJe** [comment]
glibc no. A smaller libc, more tailored for that kind of thing (i.e., musl), yes. But that's usually done only for a single binary, the overhead for a project such as this one would be *massive*.

And our buildsystem is very much not made to do that.

So, nope, not a solution ;).

**Comment by houqp** [comment]
@frankyifei what model of kindle do you have? I can help add it to https://github.com/koreader/koxtoolchain so you don't have to use the ubuntu one.

**Comment by NiLuJe** [comment]
If you get me a detailed look at what the symbol actually pulled looks like, I might be able to tell you which flag to use to work around it ;).

FWIW, I usually use this one-liner to detect issues like that, before taking a detailed look manually:

```for file in $(find . -type f) ; do ; [[ "$(file -b "${file}")" == ELF* ]] && readelf -V "${file}" | grep -q "GLIBC_2.[6-9]" && echo "Compat issue in ${file} . . ." ; done```

Assuming a Kindle running >= 5.6.5, so flagging anything > GLIBC_2.19

**Comment by NiLuJe** [comment]
Since you mention liblept, this might be because of the recent bump, and might actually also affect some of the build-bots...

I don't have the VM I used the last time to replicate the buildbot env, so...

**Comment by frankyifei** [comment]
I have a kindle pw1.

**Comment by NiLuJe** [comment]
Then you've got an extra check to do ;p.

```
for file in $(find . -type f) ; do ; [[ "$(file -b "${file}")" == ELF* ]] && readelf -V "${file}" | grep -q "GLIBC_2.1[3-9]" && echo "Compat issue in ${file} . . ." ; done
```

**Comment by NiLuJe** [comment]
Once you've identified which libraries are affected with those two commands, get me the specific symbols involved with a tailored readelf call, i.e.:

```
readelf --dyn-syms libs/liblept.so.3 | grep GLIBC_2.22
```

**Comment by houqp** [comment]
alright, i will leave it to @NiLuJe then, he might be able to come up with some build flag magic to make it work for the TC in new ubuntu. Another thing to check is to see if the new TC works for any old version before https://github.com/koreader/koreader/commit/5e8c045fcb216c3dcff3918544ddc5d8e6eb830a.

**Comment by frankyifei** [comment]
@NiLuJe , just run these two commands? I got an error
@houqp , I'll test now.

**Comment by NiLuJe** [comment]
And remind me to double-check the buildbot releases once one made after that commit hits the OTA server... ;).

**Comment by NiLuJe** [comment]
@frankyifei: Agh, crap, it's my zsh syntax. I'll update my two posts (basically, remove the semicolons after a do).

**Comment by frankyifei** [comment]
first command returned nothing and second give this `Compat issue in ./liblept.so.4 . . .`

**Comment by NiLuJe** [comment]
@frankyifei : Great, that's a good start :). Now for the tailored readelf call on that lib :).

**Comment by frankyifei** [comment]
`15: 00000000     0 FUNC    GLOBAL DEFAULT  UND fmemopen@GLIBC_2.22 (5)`

**Comment by frankyifei** [comment]
and this 
` 3: 00000000     0 FUNC    GLOBAL DEFAULT  UND calloc@GLIBC_2.4 (2)
     5: 00000000     0 FUNC    GLOBAL DEFAULT  UND strstr@GLIBC_2.4 (2)
     6: 00000000     0 FUNC    GLOBAL DEFAULT  UND raise@GLIBC_2.4 (2)
     8: 00000000     0 FUNC    GLOBAL DEFAULT  UND log@GLIBC_2.4 (4)
    10: 00000000     0 FUNC    GLOBAL DEFAULT  UND strcmp@GLIBC_2.4 (2)
    11: 00000000     0 FUNC    WEAK   DEFAULT  UND __cxa_finalize@GLIBC_2.4 (2)
    13: 00000000     0 FUNC    GLOBAL DEFAULT  UND strtol@GLIBC_2.4 (2)
    17: 00000000     0 FUNC    GLOBAL DEFAULT  UND fopen@GLIBC_2.4 (2)
    18: 00000000     0 FUNC    GLOBAL DEFAULT  UND tanf@GLIBC_2.4 (4)
    20: 00000000     0 FUNC    GLOBAL DEFAULT  UND longjmp@GLIBC_2.4 (2)
    21: 00000000     0 FUNC    GLOBAL DEFAULT  UND _setjmp@GLIBC_2.4 (2)
    22: 00000000     0 FUNC    GLOBAL DEFAULT  UND memmove@GLIBC_2.4 (2)
    24: 00000000     0 FUNC    GLOBAL DEFAULT  UND free@GLIBC_2.4 (2)
    26: 00000000     0 FUNC    GLOBAL DEFAULT  UND fgets@GLIBC_2.4 (2)
    28: 00000000     0 FUNC    GLOBAL DEFAULT  UND memcpy@GLIBC_2.4 (2)
    31: 00000000     0 FUNC    GLOBAL DEFAULT  UND time@GLIBC_2.4 (2)
    32: 00000000     0 FUNC    GLOBAL DEFAULT  UND pow@GLIBC_2.4 (4)
    33: 00000000     0 FUNC    GLOBAL DEFAULT  UND _IO_getc@GLIBC_2.4 (2)
    34: 00000000     0 FUNC    GLOBAL DEFAULT  UND ftell@GLIBC_2.4 (2)
    36: 00000000     0 FUNC    GLOBAL DEFAULT  UND sincos@GLIBC_2.4 (4)
    38: 00000000     0 FUNC    GLOBAL DEFAULT  UND rewind@GLIBC_2.4 (2)
    39: 00000000     0 FUNC    GLOBAL DEFAULT  UND ceilf@GLIBC_2.4 (4)
    41: 00000000     0 FUNC    GLOBAL DEFAULT  UND __xstat@GLIBC_2.4 (2)
    42: 00000000     0 FUNC    GLOBAL DEFAULT  UND sincosf@GLIBC_2.4 (4)
    43: 00000000     0 OBJECT  GLOBAL DEFAULT  UND stderr@GLIBC_2.4 (2)
    45: 00000000     0 FUNC    GLOBAL DEFAULT  UND fwrite@GLIBC_2.4 (2)
    46: 00000000     0 FUNC    GLOBAL DEFAULT  UND tmpfile@GLIBC_2.4 (2)
    47: 00000000     0 FUNC    GLOBAL DEFAULT  UND strcpy@GLIBC_2.4 (2)
    48: 00000000     0 FUNC    GLOBAL DEFAULT  UND gettimeofday@GLIBC_2.4 (2)
    50: 00000000     0 FUNC    GLOBAL DEFAULT  UND fread@GLIBC_2.4 (2)
    52: 00000000     0 FUNC    GLOBAL DEFAULT  UND opendir@GLIBC_2.4 (2)
    53: 00000000     0 FUNC    GLOBAL DEFAULT  UND getenv@GLIBC_2.4 (2)
    54: 00000000     0 FUNC    GLOBAL DEFAULT  UND powf@GLIBC_2.4 (4)
    55: 00000000     0 FUNC    GLOBAL DEFAULT  UND malloc@GLIBC_2.4 (2)
    57: 00000000     0 FUNC    GLOBAL DEFAULT  UND system@GLIBC_2.4 (2)
    58: 00000000     0 FUNC    GLOBAL DEFAULT  UND strftime@GLIBC_2.4 (2)
    60: 00000000     0 FUNC    GLOBAL DEFAULT  UND localtime@GLIBC_2.4 (2)
    61: 00000000     0 FUNC    GLOBAL DEFAULT  UND floorf@GLIBC_2.4 (4)
    63: 00000000     0 FUNC    GLOBAL DEFAULT  UND __ctype_b_loc@GLIBC_2.4 (2)
    64: 00000000     0 FUNC    GLOBAL DEFAULT  UND getcwd@GLIBC_2.4 (2)
    65: 00000000     0 FUNC    GLOBAL DEFAULT  UND getpid@GLIBC_2.4 (2)
    66: 00000000     0 FUNC    GLOBAL DEFAULT  UND sin@GLIBC_2.4 (4)
    69: 00000000     0 FUNC    GLOBAL DEFAULT  UND strlen@GLIBC_2.4 (2)
    71: 00000000     0 FUNC    GLOBAL DEFAULT  UND strchr@GLIBC_2.4 (2)
    73: 00000000     0 FUNC    GLOBAL DEFAULT  UND srand@GLIBC_2.4 (2)
    74: 00000000     0 FUNC    GLOBAL DEFAULT  UND fprintf@GLIBC_2.4 (2)
    76: 00000000     0 FUNC    GLOBAL DEFAULT  UND snprintf@GLIBC_2.4 (2)
    78: 00000000     0 FUNC    GLOBAL DEFAULT  UND mkdir@GLIBC_2.4 (2)
    80: 00000000     0 FUNC    GLOBAL DEFAULT  UND tan@GLIBC_2.4 (4)
    82: 00000000     0 FUNC    GLOBAL DEFAULT  UND memset@GLIBC_2.4 (2)
    83: 00000000     0 FUNC    GLOBAL DEFAULT  UND strncpy@GLIBC_2.4 (2)
    86: 00000000     0 FUNC    GLOBAL DEFAULT  UND logf@GLIBC_2.4 (4)
    88: 00000000     0 FUNC    GLOBAL DEFAULT  UND fgetc@GLIBC_2.4 (2)
    89: 00000000     0 FUNC    GLOBAL DEFAULT  UND strtod@GLIBC_2.4 (2)
    91: 00000000     0 FUNC    GLOBAL DEFAULT  UND atan@GLIBC_2.4 (4)
    93: 00000000     0 FUNC    GLOBAL DEFAULT  UND ceil@GLIBC_2.4 (4)
    94: 00000000     0 FUNC    GLOBAL DEFAULT  UND rand@GLIBC_2.4 (2)
    95: 00000000     0 FUNC    GLOBAL DEFAULT  UND fclose@GLIBC_2.4 (2)
    98: 00000000     0 FUNC    GLOBAL DEFAULT  UND expf@GLIBC_2.4 (4)
    99: 00000000     0 FUNC    GLOBAL DEFAULT  UND readdir@GLIBC_2.4 (2)
   100: 00000000     0 FUNC    GLOBAL DEFAULT  UND strrchr@GLIBC_2.4 (2)
   103: 00000000     0 FUNC    GLOBAL DEFAULT  UND fputc@GLIBC_2.4 (2)
   105: 00000000     0 FUNC    GLOBAL DEFAULT  UND open_memstream@GLIBC_2.4 (2)
   106: 00000000     0 FUNC    GLOBAL DEFAULT  UND sprintf@GLIBC_2.4 (2)
   108: 00000000     0 FUNC    GLOBAL DEFAULT  UND remove@GLIBC_2.4 (2)
   110: 00000000     0 FUNC    GLOBAL DEFAULT  UND getrusage@GLIBC_2.4 (2)
   112: 00000000     0 FUNC    GLOBAL DEFAULT  UND fseek@GLIBC_2.4 (2)
   113: 00000000     0 FUNC    GLOBAL DEFAULT  UND strncat@GLIBC_2.4 (2)
   115: 00000000     0 FUNC    GLOBAL DEFAULT  UND fputs@GLIBC_2.4 (2)
   116: 00000000     0 FUNC    GLOBAL DEFAULT  UND strncmp@GLIBC_2.4 (2)
   120: 00000000     0 FUNC    GLOBAL DEFAULT  UND closedir@GLIBC_2.4 (2)
   121: 00000000     0 FUNC    GLOBAL DEFAULT  UND atan2@GLIBC_2.4 (4)
`

**Comment by NiLuJe** [comment]
Yay. Ironically enough, it's a side-effect of one of our workarounds.

I'll see if it can be inhibited on leptonica's side...

**Comment by NiLuJe** [comment]
We can disable it in leptonica's side, but, according to the manpage... 

```
fmemopen() was already available in glibc 1.0.x.
```

So, going to dig into glibc changelogs to check why/when it was updated...

In the meantime, exporting ```ac_cv_func_fmemopen=no``` before configure for leptonica should work that around (leptonica instead uses temporary files, which might not be optimal, hence my digging into it).

**Comment by NiLuJe** [comment]
Yup...

https://sourceware.org/git/?p=glibc.git;a=commit;h=fdb7d390dd0d96e4a8239c46f3aa64598b90842b

**Comment by NiLuJe** [comment]
Does an ugly ```-Dfmemopen=__old_fmemopen``` in leptonica's CFLAGS do the job?

**Comment by NiLuJe** [comment]
Making sure we an older version of a symbol of a function which hasn't changed its signature is a hairy mess.

One solution is exposed [here](https://stackoverflow.com/questions/2856438/how-can-i-link-to-a-specific-glibc-version), but it requires code changes. And I vaguely recall not being able to reliably make it work.

**Comment by NiLuJe** [comment]
TL;DR: None of the two workarounds exposed are terribly good. Better build a proper TC targeting the right stuff directly ;).

**Comment by houqp** [comment]
i don't know, but `-Dfmemopen=__old_fmemopen` sounds reasonable to me ;P 

I am testing kindl5 tc on my laptop right now. If everything works, I will push the new config to koxtc repo.

**Comment by NiLuJe** [comment]
@houqp: See my edit, I don't think it will, since there's no public definition of __old_fmemopen, it's internal to glibc and is only exposed as a compatibility versioned symbol :/.

**Comment by NiLuJe** [comment]
@houqp: I can confirm that my K5 KOReader builds work over here, so ping me if you hit a snag ;).

**Comment by frankyifei** [comment]
why is that on ubuntu 15.10 (with glibc 2.21)it can produce working code for kpw1 (glibc 2.10)?

**Comment by houqp** [comment]
As @NiLuJe pointed out, the commit https://sourceware.org/git/?p=glibc.git;a=commit;h=fdb7d390dd0d96e4a8239c46f3aa64598b90842b was added in 2.22. So 2.21 will not have this issue. I just pushed the kindle5 platform support to the koxtc repo. Can you give that a try?

**Comment by NiLuJe** [comment]
Because we have workarounds in place for it ;). AFAIR, the buildbot is
using an Ubuntu TC. I had to match that in a VM to double-check those
tricks some time ago, c.f. the comments in Makefile.defs

And we did not hit such a tricky issue, that was a plus ;).
On May 16, 2016 7:26 AM, "Yifei(Frank) ZHU" <notifications@github.com>
wrote:

why is that on ubuntu 15.10 (with glibc 2.21)it can produce working code
for kpw1 (glibc 2.10)?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/koreader/koreader-base/issues/421#issuecomment-219351439>

**Comment by houqp** [comment]
The buildbot runs on trusty fwiw.

**Comment by NiLuJe** [comment]
Ah, great, thanks, so we're safe from that one ;).

**Comment by NiLuJe** [comment]
82511d5 appears to do the trick on my end, and is slightly less messy than I feared ;).

**Comment by NiLuJe** [comment]
... And of course something went wrong ;D.

The default symbol name in glibc < 2.22 was ```fmemopen@@GLIBC_2.4```, so, of course, if applied unilaterally, this breaks linking on those systems.

Yay.

**Comment by NiLuJe** [comment]
That was fun. (not).

Currently checking that this works as intended...

**Comment by lgeek** [comment]
I'm not saying this is not a good solution or that it should be changed. However, shipping a standard libc shared object with koreader, to be used across all targets seems to be another option that hasn't been discussed. It might be useful to keep it in mind as an option if diverging glibc versions become an issue. Any thoughts?

**Comment by NiLuJe** [comment]
Bound to cause problems, IIRC. (In the same vein as linking glibc statically being unsupported).

**Comment by houqp** [comment]
@NiLuJe if it's just one shared glibc, then it take less than 3m space right? However, I can see a lot of work is needed to get it right. I think if someone can come up with a creative way to make it easy to maintain, then it's not a bad idea :)

**Comment by NiLuJe** [comment]
@houqp : Glibc is a *tiny* bit more than libc.so.6 ;).

```
equery f glibc | ag '^/usr/lib64/' | wc -l
305
```

Most importantly, the dynamic loader used *has* to match the libc used, and everything in the stack has to follow that (which is why I mentioned @knc1 experiment, because it does exactly this).

So, yeah, I feel like that's a lot of overengineering to fix an issue which has much better & easier workarounds ;).
